sparse/coo.py

```diff
@@ -2034,113 +1988,127 @@ def _grouped_reduce(x, groups, method, **kwargs):
+        don't have broadcastable shapes.
+
+    See Also
+    --------
+    :obj:`numpy.ufunc` : A similar Numpy construct. Note that any :code:`ufunc` can be used
+        as the :code:`func` input to this function.
+    """
+    # Because we need to mutate args.
+    args = list(args)
+    posargs = []
+    pos = []
+    for i in range(len(args)):
+        if isinstance(args[i], scipy.sparse.spmatrix):
+            args[i] = COO.from_scipy_sparse(args[i])
+
+        if isscalar(args[i]) or (isinstance(args[i], np.ndarray)
+                                 and not args[i].shape):
+            # Faster and more reliable to pass ()-shaped ndarrays as scalars.
+            if isinstance(args[i], np.ndarray):
+                args[i] = args[i][()]
+
+            # The -scalars factor is there because we need to account for already
+            # added scalars in the function.
+            pos.append(i)
+            posargs.append(args[i])
+        elif isinstance(args[i], SparseArray) and not isinstance(args[i], COO):
+            args[i] = COO(args[i])
+        elif not isinstance(args[i], COO):
             raise ValueError("Performing this operation would produce "
                              "a dense result: %s" % str(func))
 
-    self_shape, other_shape = self.shape, other.shape
-
-    result_shape = _get_broadcast_shape(self_shape, other_shape)
-    self_params = _get_broadcast_parameters(self.shape, result_shape)
-    other_params = _get_broadcast_parameters(other.shape, result_shape)
-    combined_params = [p1 and p2 for p1, p2 in zip(self_params, other_params)]
-    self_reduce_params = combined_params[-self.ndim:]
-    other_reduce_params = combined_params[-other.ndim:]
-
-    self.sum_duplicates()  # TODO: document side-effect or make copy
-    other.sum_duplicates()  # TODO: document side-effect or make copy
-
-    self_coords = self.coords
-    self_data = self.data
-
-    self_reduced_coords, self_reduced_shape = \
-        _get_reduced_coords(self_coords, self_shape,
-                            self_reduce_params)
-    self_reduced_linear = _linear_loc(self_reduced_coords, self_reduced_shape)
-    i = np.argsort(self_reduced_linear)
-    self_reduced_linear = self_reduced_linear[i]
-    self_coords = self_coords[:, i]
-    self_data = self_data[i]
-
-    # Store coords
-    other_coords = other.coords
-    other_data = other.data
-
-    other_reduced_coords, other_reduced_shape = \
-        _get_reduced_coords(other_coords, other_shape,
-                            other_reduce_params)
-    other_reduced_linear = _linear_loc(other_reduced_coords, other_reduced_shape)
-    i = np.argsort(other_reduced_linear)
-    other_reduced_linear = other_reduced_linear[i]
-    other_coords = other_coords[:, i]
-    other_data = other_data[i]
+    # Filter out scalars as they are 'baked' into the function.
+    func = _posarg_partial(func, pos, posargs)
+    args = list(filter(lambda arg: not isscalar(arg), args))
```

Comment:
You might consider `toolz.remove` here

Reply:
I'd prefer not to introduce a dependency for something as simple as this.